ompi/mca/coll/base/coll_base_alltoall.c

```diff
@@ -297,42 +376,48 @@ int ompi_coll_base_alltoall_intra_bruck(const void *sbuf, size_t scount,
     }
 
     /* perform communication step */
-    for (distance = 1; distance < size; distance<<=1) {
-
-        sendto = (rank + distance) % size;
-        recvfrom = (rank - distance + size) % size;
+    for (distance = 1; distance < size; distance *= radix) {
+        for (i = 1; i < radix; i++) {
 
-        new_ddt = ompi_datatype_create((1 + size/distance) * (2 + rdtype->super.desc.used));
-
-        /* Create datatype describing data sent/received */
-        for (i = distance; i < size; i += 2*distance) {
-            int nblocks = distance;
-            if (i + distance >= size) {
-                nblocks = size - i;
+            if (distance * i >= size) {
+                break;
             }
-            ompi_datatype_add(new_ddt, rdtype, rcount * nblocks,
-                              i * rcount * rext, rext);
-        }
 
-        /* Commit the new datatype */
-        err = ompi_datatype_commit(&new_ddt);
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl;  }
+            sendto = (rank + distance * i) % size;
+            recvfrom = (rank - distance * i + size) % size;
 
-        /* Sendreceive */
-        err = ompi_coll_base_sendrecv ( tmpbuf, 1, new_ddt, sendto,
-                                         MCA_COLL_BASE_TAG_ALLTOALL,
-                                         rbuf, 1, new_ddt, recvfrom,
-                                         MCA_COLL_BASE_TAG_ALLTOALL,
-                                         comm, MPI_STATUS_IGNORE, rank );
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl; }
+            new_ddt = ompi_datatype_create((1 + size/distance) * (2 + rdtype->super.desc.used));
 
-        /* Copy back new data from recvbuf to tmpbuf */
-        err = ompi_datatype_copy_content_same_ddt(new_ddt, 1,tmpbuf, (char *) rbuf);
-        if (err < 0) { line = __LINE__; err = -1; goto err_hndl;  }
+            /* Create datatype describing data sent/received */
+            for (j = i * distance; j < size; j += radix * distance) {
+                int nblocks = distance;
+                if (j + distance >= size) {
+                    nblocks = size - j;
+                }
+                ompi_datatype_add(new_ddt, rdtype, rcount * nblocks,
+                                  j * rcount * rext, rext);
+            }
 
-        /* free ddt */
-        err = ompi_datatype_destroy(&new_ddt);
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl;  }
+            /* Commit the new datatype */
+            err = ompi_datatype_commit(&new_ddt);
```

Comment:
I think the real reason you are not seeing benefits by using nonblocking communications is due to the cost of the datatype creation. During this time, the posted communications will not be able to progress, which means the bandwidth is wasted until all datatype are build and you reach the waitall part.

Reply:
Sorry I missed this comment from earlier. We profiled the application and found that datatype creation time was insignificant, compared to the memcpy and wait time.